crates/test/macros/src/lib.rs

```diff
@@ -0,0 +1,170 @@
+    let test_names = tests.iter().map(|test| &test.sig.ident);
+
+    Ok(quote!(
+        #[cfg(test)]
+        pub(crate) mod #ident {
+            #(#untouched_tokens)*
+
+            #(#tests)*
+
+            pub(crate) fn __test_list() -> impl Iterator<Item = fn()> {
+                const TEST_LIST: [fn(); #test_count] = [#(#test_names), *];
+                TEST_LIST.into_iter()
+            }
+        }
+    )
+    .into())
+}
+
+#[proc_macro]
+pub fn tests_runner(args: TokenStream) -> TokenStream {
+    match tests_runner_impl(args) {
+        Ok(ts) => ts,
+        Err(e) => e.to_compile_error().into(),
+    }
+}
+
+fn tests_runner_impl(args: TokenStream) -> parse::Result<TokenStream> {
+    let args: syn::ExprArray = syn::parse(args)?;
+
+    let test_suites = args
+        .elems
+        .into_iter()
+        .map(|attr| quote!(let ret = ret.chain(#attr::__test_list());))
+        .collect::<Vec<_>>();
+
+    Ok(quote!(
+        #[cfg(test)]
+        mod __test_runner {
+            // Required for panic handler
+            extern crate flipperzero_rt;
+
+            // Required for allocator
+            #[cfg(feature = "alloc")]
+            extern crate flipperzero_alloc;
+
+            use core::time::Duration;
+
+            use flipperzero_rt::{entry, manifest};
+            use flipperzero_sys as sys;
+
+            use crate::{furi::thread, println};
+
+            manifest!(name = "Rust Unit Tests");
+            entry!(main);
+
+            fn test_list() -> impl Iterator<Item = fn()> {
+                let ret = ::core::iter::empty();
+                #(#test_suites)*
+                ret
+            }
+
+            // Test runner entry point
+            fn main(_args: *mut u8) -> i32 {
+                let heap_before = unsafe { sys::memmgr_get_free_heap() };
+                let cycle_counter = unsafe { sys::furi_get_tick() };
+
+                let failed = 0;
+                for test in test_list() {
+                    // When the test assertions fail, we can't catch the panics.
+                    // TODO: Replace panics with Results.
+                    test();
+                }
+
+                println!("");
+                println!("Failed tests: {}\r", failed);
+
+                // Time report
+                println!(
+                    "Consumed: {} ms\r",
+                    unsafe { sys::furi_get_tick() } - cycle_counter
+                );
+
+                // Wait for tested services and apps to deallocate memory
+                thread::sleep(Duration::from_millis(200));
+                let heap_after = unsafe { sys::memmgr_get_free_heap() };
+                println!("Leaked: {}\r", heap_before - heap_after);
+
+                // Final Report
+                if failed == 0 {
+                    println!("Status: PASSED\r");
+                    0
+                } else {
+                    println!("Status: FAILED\r");
+                    1
+                }
```

Comment:
Currently the test runner prints the same information as the [`flipperzero-firmware` unit tests](https://github.com/flipperdevices/flipperzero-firmware/blob/0.78.1/documentation/UnitTests.md) in the device logs. However, for the purpose of `cargo test` we need to extract this information back to the host. I currently believe this means we'll need to save the results to a file and then download it before cleaning up the device.

Reply:
Saving the results to a file is what I am doing now.